src/mrpro/operators/WaveletOp.py

```diff
@@ -0,0 +1,359 @@
+                    'entries for all considered dimensions.'
+                )
+
+            # size of wavelets
+            wavelet_length = torch.as_tensor((Wavelet(wavelet_name).dec_len,) * len(domain_shape))
+
+            # calculate shape of wavelet coefficients at each level
+            current_shape = torch.as_tensor(domain_shape)
+
+            # if level is None, select the highest possible level.
+            # raise error/warnings if level is not possible or lead to boundary effects
+            verified_level = _check_level(domain_shape, wavelet_length, level)
+
+            if verified_level == 0:
+                self.coefficients_shape = [domain_shape]
+            else:
+                self.coefficients_shape = []
+                for _ in range(verified_level):
+                    # Add padding
+                    current_shape = (current_shape / 2).ceil() + wavelet_length // 2 - 1
+                    self.coefficients_shape.extend(
+                        [tuple(current_shape.to(dtype=torch.int64))] * self.n_wavelet_directions
+                    )
+
+                self.coefficients_shape = self.coefficients_shape[::-1]
+                self.coefficients_shape.insert(0, self.coefficients_shape[0])  # shape of a/aa/aaa term
+
+    def forward(self, x: torch.Tensor) -> tuple[torch.Tensor,]:
+        """Calculate wavelet coefficients from (image) data.
+
+        Parameters
+        ----------
+        x
+            (Image) data
+
+        Returns
+        -------
+            Wavelet coefficients stacked along one dimension
+
+        Raises
+        ------
+        ValueError
+            If the dimensions along which wavelets are to be calculated are not unique.
+        """
+        # normalize axes to allow negative indexing in input
+        dim = tuple(d % x.ndim for d in self.dim)
+        if len(dim) != len(set(dim)):
+            raise ValueError(f'Axis must be unique. Normalized axis are {dim}')
+
+        # swapping the last axes and the axes to calculate wavelets of
+        x = torch.moveaxis(x, dim, list(range(-len(self.dim), 0)))
+
+        # the ptwt functions work only for real data, thus we handle complex inputs as an additional channel
+        x_real = torch.view_as_real(x).moveaxis(-1, 0) if x.is_complex() else x
+
+        if len(self.dim) == 1:
+            coeffs1 = wavedec(x_real, self.wavelet_name, level=self.level, mode='zero', axis=-1)
+            coefficients_list = self._format_coeffs_1d(coeffs1)
+        elif len(self.dim) == 2:
+            coeffs2 = wavedec2(x_real, self.wavelet_name, level=self.level, mode='zero', axes=(-2, -1))
+            coefficients_list = self._format_coeffs_2d(coeffs2)
+        elif len(self.dim) == 3:
+            coeffs3 = wavedec3(x_real, self.wavelet_name, level=self.level, mode='zero', axes=(-3, -2, -1))
+            coefficients_list = self._format_coeffs_3d(coeffs3)
+
```

Comment:
else: raise error?

dim is public and could be changed. also, adding this check here does not cost anything...

Reply:
why does this pass mypy?
 coefficients_list could be uninitialized and the self._coeff_to_stacked_tensor would fail